### PR TITLE
If there is just a single root module, document it in index.html

### PR DESCRIPTION
**Pull request by not-my-profile**
I think most of the time `pydoctor` is just invoked with a single module. In that case it's nicer if `index.html` immediately shows the module documentation instead of just a page linking the module.

### CODE REVIEW

File: pydoctor/templatewriter/writer.py [hunk 1]
```diff
@@ -97,7 +97,7 @@ def _writeDocsFor(self, ob: model.Documentable) -> None:
             if self.dry_run:
                 self.total_pages += 1
             else:
-                with self.build_directory.joinpath(f'{ob.fullName()}.html').open('wb') as fobj:
+                with self.build_directory.joinpath(ob.url.split('#')[0]).open('wb') as fobj:
```
Comment (tristanlatr) [inline pydoctor/templatewriter/writer.py:100]: Why this change ?

**Comment by tristanlatr** [comment]
@not-my-profile could you stop force pushing from the time we start reviewing your code? you can use regular merge instead of rebase and then we can squash everything together at the end. It’s a bit hard to follow your changes.

**Comment by not-my-profile** [comment]
I don't actually want you to squash my commits. E.g. for this PR i intentionally split the refactor into a separate commit.

I think you should wait with reviewing until the CI succeeded ... if it fails I'll force push a fix.

**Comment by tristanlatr** [comment]
> I don't actually want you to squash my commits. 

Ok.

Force pushing messes with the review. I’ll wait for the tests to be green to review. 

But if there are changes to be done after the first review, i’d like those changes not to be forced pushed. Such that I can review only the new changes.

**Comment by tristanlatr** [comment]
Does it breaks links that were pointing to the single root ? Maybe, in case of single root, index.html should only be a redirect to the module page (url unchanged). What do you think ?

**Comment by not-my-profile** [comment]
Right, I'll mark my PRs as drafts in the future until they're ready for review. Just FYI I always try to split up my changes into atomic commits to make them easier to review. So I recommend reviewing my changes commit by commit.

File: pydoctor/driver.py [hunk 2]
```diff
@@ -489,6 +489,12 @@ def p(msg: str) -> None:
                 for fn in sorted(system.docstring_syntax_errors):
                     p('    '+fn)
 
+            if len(system.root_names) == 1:
+                # If there is just a single root module it is written to index.html to produce nicer URLs.
+                # To not break old links we also create a symlink from the full module name to the index.html
+                # file. This is also good for consistency: every module is accessible by <full module name>.html
+                (build_directory / (list(system.root_names)[0] + '.html')).symlink_to('index.html')
```
Comment (not-my-profile) [inline pydoctor/driver.py:496]: I just added a check to create a symlink so that old links outside of pydoctor don't break.

File: pydoctor/templatewriter/writer.py [hunk 3]
```diff
@@ -82,22 +82,30 @@ def writeIndividualFiles(self, obs: Iterable[model.Documentable]) -> None:
...
                 flattenToFile(fobj, page)
             system.msg('html', "took %fs"%(time.time() - T), wantsnl=False)
 
+        if len(system.root_names) == 1:
+            # If there is just a single root module it is written to index.html to produce nicer URLs.
+            # To not break old links we also create a symlink from the full module name to the index.html
+            # file. This is also good for consistency: every module is accessible by <full module name>.html
+            root_module_path = (self.build_directory / (list(system.root_names)[0] + '.html'))
+            root_module_path.unlink()
+            root_module_path.symlink_to('index.html')
```
Comment (not-my-profile) [inline pydoctor/templatewriter/writer.py:99]: I just added a check to create a symlink so that old links outside of pydoctor don't break. I initially tried adding the check in driver but then the symlink creation fails because some tests use an `InMemoryWriter`. So having the check in the `TemplateWriter` class makes more sense.

**Comment by not-my-profile** [comment]
Ok this is now ready for review :) Btw @tristanlatr did you get my mail?

File: pydoctor/templatewriter/writer.py [hunk 4]
```diff
@@ -82,22 +82,34 @@ def writeIndividualFiles(self, obs: Iterable[model.Documentable]) -> None:
...
+
     def _writeDocsFor(self, ob: model.Documentable) -> None:
         if not ob.isVisible:
             return
         if ob.documentation_location is model.DocLocation.OWN_PAGE:
             if self.dry_run:
                 self.total_pages += 1
             else:
-                with self.build_directory.joinpath(f'{ob.fullName()}.html').open('wb') as fobj:
+                with self.build_directory.joinpath(ob.url.split('#')[0]).open('wb') as fobj:
```
Comment (tristanlatr) [inline pydoctor/templatewriter/writer.py:112]: Not sure we absolutely need to split on `#`. Since `ob.documentation_location is model.DocLocation.OWN_PAGE`, `page_object` will return `self` and nothing will be added after the filename in the URL. 

I think the following is OK:

```suggestion
                # Documentable.page_object.url will alway be a plain filename.
                with self.build_directory.joinpath(ob.page_object.url).open('wb') as fobj:
```

**Review by tristanlatr — COMMENTED** [review]
Only a minor comment. 

I really like this feature. Thanks

File: pydoctor/test/test_templatewriter.py [hunk 5]
```diff
@@ -104,7 +104,7 @@ def test_basic_package(tmp_path: Path) -> None:
         if '#' in url:
             url = url[:url.find('#')]
         assert (tmp_path / url).is_file()
-    with open(tmp_path / 'basic.html', encoding='utf-8') as f:
+    with open(tmp_path / 'index.html', encoding='utf-8') as f:
```
Comment (tristanlatr) [inline pydoctor/test/test_templatewriter.py:107]: Isn’t this supposed to still work with “basic.html” ?

File: pydoctor/templatewriter/writer.py (hunk 4 above)
Comment (not-my-profile) [reply pydoctor/templatewriter/writer.py:112]: I think splitting here is cleaner, because otherwise there is an undocumented constraint on the `url` method. If you really want to avoid that split we'd need to document that in the `url` method that it should not set an anchor for "page objects".

Comment (tristanlatr) [reply pydoctor/templatewriter/writer.py:112]: Yes, to mt it's pretty clear the that the url method that it should not set an anchor for "page objects".

File: pydoctor/test/test_templatewriter.py (hunk 5 above)
Comment (not-my-profile) [reply pydoctor/test/test_templatewriter.py]: Yes good point. I made this change initially since I didn't create the symlinks initially. I have remove this change.

File: pydoctor/templatewriter/writer.py (hunk 4 above)
Comment (not-my-profile) [reply pydoctor/templatewriter/writer.py:112]: Right. I still prefer such assumptions to be documented. I feel like not depending on such behavior is even better.

Comment (tristanlatr) [reply pydoctor/templatewriter/writer.py:112]: You could quickly add a note to the url property docstring saying that page object should always return a plain relative file name, without anchor.

Comment (not-my-profile) [reply pydoctor/templatewriter/writer.py]: Done.

**Review by tristanlatr — APPROVED** [review]
Thanks
